specification/k3po.control/SPEC.md

```diff
@@ -0,0 +1,258 @@
+>                      LF
+>                      [ message-body LF ]
+
+command-line      = command LF
+
+command              = token
+
+header-field = field-name ":" field-value
+
+LF, message-body, field-name, field-value, and token are defined in [HTTP RFC](https://tools.ietf.org/html/rfc7230).
+
+If the "content-length" header is specified, then a message body of the specified length must be added.
+
+Headers that are not understood / processed by a command must be ignored.
+
+
+### Flow
+
+The communication between the test framework and the driver is a conversation - the test framework sends a command, the driver responds with an
+event sent asynchronously when the action is finished. The only exception is the FINISHED event - the START command will have two events as a 
+response - STARTED when the execution of the scripts is started and FINISHED when the execution of the scripts is finished.
+
+Throughout the flow description bellow, **an error** represents an ERROR event sent by the driver.
+
+- The test framework initiates the conversation with a **PREPARE** command. This command will make the k3po server load and prepare a set
+of scripts for execution. The scripts are not yet executed, just loaded from disk and parsed. If there are ACCEPT instructions in the scripts,
+a bound on corresponding ports is performed. If the command is successful, the driver sends back a PREPARED event. In case of failure, an
+ERROR event is sent.
+
+- If the PREPARED command is successful, a **START** command may be sent by the test framework. This command will start the execution of the
+ clients in the scripts that were sent in the PREPARE command. If the command is successful, the driver sends back a STARTED event. This 
+ command will also trigger the sending of a FINISHED event whenever the scripts finished execution on the k3po server.
+ 
+- If there is no PREPARE command send before the START command or if the PREPARE command was not successful an error with a message "Script has not
+ been prepared or is still preparing" must occur. 
+ 
+- If the START command was already sent, an error with a message "Script has already been started" must occur.
+
+- Any time after the PREPARE command and before the DISPOSE command, the test framework can send an **AWAIT** command for a barrier existing 
+in the scripts. The command receives a response with a NOTIFIED event only when the corresponding barrier is notified. In case the barrier 
+does not exist, an error must be returned.
+
+- Any time after the PREPARE command *and before DISPOSE?*, the test framework can send a **NOTIFY** command for a barrier existing in the scripts.
+ The command notifies the barrier and a response with a NOTIFIED event is received. In case the barrier does not exist, an error must be returned.
+
+- Any time after the PREPARE command *and before DISPOSE?*, the test framework can send an **ABORT** command. The driver must abort the current 
```

Comment:
I just understood the way responses are given to ABORT - the FINISHED event is sent only once per test. In case the script is finished (and FINISHED sent), any ABORT command will have no response. If the script is not finished, the first ABORT will trigger a FINISHED event, subsequent ABORT commands will have no response. I have updated the flow accordingly.